aviatrix/resource_aviatrix_gateway.go

```diff
@@ -2501,31 +2502,71 @@ func resourceAviatrixGatewayUpdate(d *schema.ResourceData, meta interface{}) err
 		}
 	}
 
-	if d.HasChanges("software_version", "image_version") {
-		swVersion := d.Get("software_version").(string)
-		imageVersion := d.Get("image_version").(string)
-		gw := &goaviatrix.Gateway{
-			GwName:          d.Get("gw_name").(string),
-			SoftwareVersion: swVersion,
-			ImageVersion:    imageVersion,
-		}
-		err := client.UpgradeGateway(gw)
-		if err != nil {
-			return fmt.Errorf("could not upgrade primary gateway during update image_version=%s software_version=%s: %v", gw.ImageVersion, gw.SoftwareVersion, err)
-		}
-	}
-	if haEnabled && d.HasChanges("peering_ha_image_version", "peering_ha_software_version") {
-		haSwVersion := d.Get("peering_ha_software_version").(string)
-		haImageVersion := d.Get("peering_ha_image_version").(string)
-		if haSwVersion != "" || haImageVersion != "" {
+	primaryHasVersionChange := d.HasChanges("software_version", "image_version")
+	haHasVersionChange := haEnabled && d.HasChanges("peering_ha_software_version", "peering_ha_image_version")
+	primaryHasImageVersionChange := d.HasChange("image_version")
+	haHasImageVersionChange := d.HasChange("peering_ha_image_version")
+	if primaryHasVersionChange || haHasVersionChange {
+		if primaryHasVersionChange && haHasVersionChange && !primaryHasImageVersionChange && !haHasImageVersionChange {
+			// Both Primary and HA have changed just their software_version
+			// so we can perform upgrade in parallel.
+			swVersion := d.Get("software_version").(string)
+			gw := &goaviatrix.Gateway{
+				GwName:          d.Get("gw_name").(string),
+				SoftwareVersion: swVersion,
+			}
+			haSwVersion := d.Get("peering_ha_software_version").(string)
 			hagw := &goaviatrix.Gateway{
-				GwName:          gateway.GwName + "-hagw",
+				GwName:          d.Get("gw_name").(string) + "-hagw",
 				SoftwareVersion: haSwVersion,
-				ImageVersion:    haImageVersion,
 			}
-			err = client.UpgradeGateway(hagw)
-			if err != nil {
-				return fmt.Errorf("could not upgrade PSF HA gateway during update image_version=%s software_version=%s: %v", hagw.ImageVersion, hagw.SoftwareVersion, err)
+			var wg sync.WaitGroup
+			wg.Add(2)
+			var primaryErr, haErr error
+			go func() {
+				primaryErr = client.UpgradeGateway(gw)
+				wg.Done()
+			}()
+			go func() {
+				haErr = client.UpgradeGateway(hagw)
+				wg.Done()
+			}()
+			wg.Wait()
+			if primaryErr != nil && haErr != nil {
+				return fmt.Errorf("could not upgrade either primary or HA gateway "+
```

Comment:
and* ?

Reply:
Changed to 'could not upgrade primary and HA..."